src/api/controllers/db-controller.ts

```diff
@@ -828,43 +929,176 @@ export async function getTxFromDataStore(
-        dbTx.contract_call_function_args
-      ).map((c, fnArgIndex) => {
-        const functionArgAbi = functionAbi.args[fnArgIndex++];
-        return {
-          hex: bufferToHexPrefixString(serializeCV(c)),
-          repr: cvToString(c),
-          name: functionArgAbi.name,
-          type: getTypeString(functionArgAbi.type),
-        };
-      });
-    }
   }
+  parsedTx.contract_call.function_signature = abiFunctionToString(functionAbi);
+  if (dbTransaction.contract_call_function_args) {
+    parsedTx.contract_call.function_args = readClarityValueArray(
+      dbTransaction.contract_call_function_args
+    ).map((c, fnArgIndex) => {
+      const functionArgAbi = functionAbi.args[fnArgIndex++];
+      return {
+        hex: bufferToHexPrefixString(serializeCV(c)),
+        repr: cvToString(c),
+        name: functionArgAbi.name,
+        type: getTypeString(functionArgAbi.type),
+      };
+    });
+  }
+  return parsedTx;
+}
+
+export async function searchTxs(
+  db: DataStore,
+  args: GetTxsArgs | GetTxsWithEventsArgs
+): Promise<FoundOrNot<Transaction | MempoolTransaction | string>[]> {
+  const minedTxs = await getTxsFromDataStore(db, args);
+
+  // filtering out mined transactions in canonical chain
+  const minedAndInCanonicalChain = minedTxs.filter(tx => tx.canonical && tx.microblock_canonical);
+
+  // filtering out tx_ids that are not in canonical chain
+  const nonCanonical = minedTxs.filter(tx => !tx.canonical && !tx.microblock_canonical);
+
+  // filtering out tx_ids that were not mined / found
+  const notMinedTransactions: string[] = args.txIds.filter(
+    txId => !minedTxs.find(minedTx => txId === minedTx.tx_id)
+  );
+
+  // finding transactions that are not mined and are not in canonical in mempool.
+  const mempoolTxs = [...nonCanonical.map(tx => tx.tx_id), ...notMinedTransactions];
+  const mempoolTxsQuery = await getMempoolTxsFromDataStore(db, {
+    txIds: mempoolTxs,
+    includeUnanchored: args.includeUnanchored,
+  });
+
+  const foundOrNotMined = minedAndInCanonicalChain.map(minedTx => {
+    return { found: true, result: minedTx };
+  });
+  const foundOrNotMempool = mempoolTxsQuery.map((mtx: any) => {
+    return { found: true, result: mtx };
+  });
+  // such transactions that are not in mempoolTxQuery but do exist in notInCanonical
+  const foundOrNotNonCanonical = nonCanonical
+    .filter(tx => mempoolTxsQuery.findIndex((mtx: any) => mtx.tx_id === tx.tx_id) < 0)
+    .map(tx => {
+      return { found: true, result: tx };
+    });
+  // all transactions that were not found anywhere
+  const foundOrNotFoundTxs = [...foundOrNotMined, ...foundOrNotMempool, ...foundOrNotNonCanonical];
+  const foundOrNotNotFoundTxs = args.txIds
+    .filter(txId => foundOrNotFoundTxs.findIndex(ftx => ftx.result.tx_id === txId) < 0)
+    .map(txId => {
+      return { found: false, result: txId };
+    });
+
+  return [...foundOrNotFoundTxs, ...foundOrNotNotFoundTxs];
```

Comment:
Seems like this should be returning txs in the order that they were requested, as in this response array should be in the same order as the request arg array.

Reply:
Done.